File: Data Structures & Algorithms/max-water-container/submission-4.js
class Solution {
    maxArea(arr) {
        let i = 0, j = arr.length - 1;
        let maxVolumn = 0;
        while(i<j){
            let width = j - i;
            let currHeight = Math.min(arr[i], arr[j]);
            let currentVolumn = width * currHeight;
            if(maxVolumn < currentVolumn) {
                maxVolumn = currentVolumn;
            }
            if( arr[i] < arr[j]) {
                i++;
            } else {
                j--;
            }
        }

        return maxVolumn;
    }
}




// class Solution {
//     /**
//      * @param {number[]} heights
//      * @return {number}
//      */
//     maxArea(heights) {
//         let i = 0;
//         let j = heights.length - 1;
//         let area = 0;

//         while(i < j) {
//             const tempArea = Math.min(heights[i], heights[j]) * (j-i);
//             area = Math.max(area, tempArea);
//             if(heights[i] <= heights[j]){
//                 i++;
//             } else {
//                 j--;
//             }
//         }
//         return area;
//     }
// }

File: Data Structures & Algorithms/trapping-rain-water/submission-4.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        if (n == 0) {
            return 0;
        }

        int[] maxLeft = new int[n];
        int[] maxRight = new int[n];

        maxLeft[0] = 0;
        for(int i=1; i<n; i++){
            maxLeft[i] = Math.max(maxLeft[i-1],height[i-1]);
        }
        maxRight[n-1]=0;
        for(int j=n-2; j>=0; j--){
            maxRight[j] = Math.max(maxRight[j+1],height[j+1]);
        }
        int resultArea = 0;
        for(int i=0; i<n; i++){
            int area = Math.min(maxLeft[i],maxRight[i]) - height[i];
            if(area>0){
                resultArea += area;
            }
        }
        return resultArea;
    }
}
